// fb-chat-support/src/changeProfile.js
'use strict';

var utils = require('../utils');
var log = require('npmlog');

module.exports = function (http, api, ctx) {
  return function changeProfile(userID, callback) {
    var cb;
    var rtPromise = new Promise(function (resolve, reject) {
      cb = error => error ? reject(error) : resolve();
    });

    if (typeof callback == 'function') cb = callback;

    var form = {
      fb_api_caller_class: 'RelayModern',
      fb_api_req_friendly_name: 'CometProfileSwitchMutation',
      variables: JSON.stringify({
        profile_id: userID
      }),
      server_timestamps: !0,
      doc_id: '4855674131145168'
    }

    http
      .post('https://www.facebook.com/api/graphql/', ctx.jar, form)
      .then(utils.saveCookies(ctx.jar))
      .then(function (res) {
        console.log(res.body);
        var html = res.body;
        var res = html ? JSON.parse(html.split(';').pop()) : null;
        if (res && (res.error || !res.data.profile_switcher_comet_login)) throw res;
        ctx.userID = userID;
        return cb();
      })
      .catch(function (err) {
        log.error('changeProfile', err);
        return cb(err);
      });
    
    return rtPromise;
  }
}
